califa/calibration/R3BCalifaCrystalCal2Cluster.cxx

```diff
@@ -138,193 +242,317 @@ InitStatus R3BCalifaCrystalCal2Cluster::ReInit()
+
+        for (Int_t j = 0; j < allCrystalVec.size(); j++)
+        {
 
-        R3BLOG(debug, "crystalId2Pos.size()=" << crystalId2Pos.size());
+            Int_t thisCryId = allCrystalVec.at(j)->GetCrystalId();
 
-        for (auto& k1 : crystalId2Pos) // k1: lower id, gamma branch?
-            if (crystalId2Pos.count(k1.first + fNbCrystalsGammaRange))
+            if (thisCryId <= fTotalCrystals && !isInside(usedCrystals, thisCryId) && !fSimulation)
             {
-                auto proton = *crystalId2Pos.find(k1.first + fNbCrystalsGammaRange);
-                // k2: higher id, proton branch
-                if (proton.second->GetEnergy() < fDRThreshold)
-                    addHit(k1.second); // gamma
-                else
-                    addHit(proton.second);
+
+                angles = fCalifaGeo->GetAngles(thisCryId) - fCalifatoTargetPos;
+
+                if (InsideClusterWindow(mother_angles, angles))
+                    addCrystal2Cluster(&cluster, allCrystalVec.at(j), "gammas", &usedCrystals, fTotalCrystals);
             }
-            else if (!crystalId2Pos.count(k1.first - fNbCrystalsGammaRange))
-                // not a hit where two ranges were hit
-                addHit(k1.second);
+        }
+
+        for (Int_t j = 0; j < allCrystalVec.size(); j++)
+        {
+
+            Int_t thisCryId = allCrystalVec.at(j)->GetCrystalId();
+
+            if (thisCryId > fTotalCrystals && !isInside(usedCrystals, thisCryId) && !fSimulation)
+            {
+
+                angles = fCalifaGeo->GetAngles(thisCryId - fTotalCrystals) - fCalifatoTargetPos;
+
+                if (InsideClusterWindow(mother_angles, angles))
+                    addCrystal2Cluster(&cluster, allCrystalVec.at(j), "proton", &usedCrystals, fTotalCrystals);
+            }
+        }
+
+        for (Int_t j = 0; j < allCrystalVec.size(); j++)
+        {
+
+            Int_t thisCryId = allCrystalVec.at(j)->GetCrystalId();
+
+            if (!isInside(usedCrystals, thisCryId) && fSimulation)
+            {
+
+                angles = fCalifaGeo->GetAngles(thisCryId) - fCalifatoTargetPos;
+
+                if (InsideClusterWindow(mother_angles, angles))
+                    addCrystal2Cluster(&cluster, allCrystalVec.at(j), "gammas", &usedCrystals, fTotalCrystals);
+            }
+        }
+
+        AddHit(
+            cluster.crystalList, cluster.energy, cluster.nf, cluster.ns, cluster.theta, cluster.phi, cluster.time, 1);
+
+        RemoveUsedCrystals(
+            usedCrystals, allCrystalVec, protonCandidatesVec, gammaCandidatesVec, saturatedCandidatesVec);
     }
-    R3BLOG(debug, "after uniquifying, we have " << unusedCrystalHits.size() << " crystal hits.");
 
-    unusedCrystalHits.sort(
-        [](R3BCalifaCrystalCalData* lhs, R3BCalifaCrystalCalData* rhs) { return lhs->GetEnergy() > rhs->GetEnergy(); });
-    uint32_t clusterId = 0;
-    while (!unusedCrystalHits.empty())
+    /* ----------- Saturation Clusters ----------- */
```

Comment:
I still do not get the point of the saturation clusters... Is it representing some other particles? Or just to remove "bad hits" due to the limitation of the non-linearity of the detectors or electronics?

Reply:
Anyway, if you can have a common sub-routine, it would be great..

Reply:
My suggestion was to handle saturation clusters by setting their energy to INFINITY. That way, any cluster for which we have an exclusive high gain (e.g. no low gain available, e.g. single range) overflow would end up being infty, which would mean that we could keep using one loop to do all the clustering. Eventually we should use time over threshold for the overflows. @latar had a look at that recently.

Reply:
This already answer on top.